Project2015To2017/ProjectConverter.cs

```diff
@@ -56,83 +56,111 @@ IProgress<string> progress
 			return Convert(target, new ConversionOptions(), preTransforms, postTransforms, progress);
 		}
 
-		public static IEnumerable<Definition.Project> Convert(
+		public static IEnumerable<Project> Convert(
 			string target,
 			ConversionOptions conversionOptions,
 			IReadOnlyList<ITransformation> preTransforms,
 			IReadOnlyList<ITransformation> postTransforms,
 			IProgress<string> progress
 		)
 		{
-			if (Path.GetExtension(target).Equals(".sln", StringComparison.OrdinalIgnoreCase))
+			var extension = Path.GetExtension(target);
+			if (string.Equals(extension, ".sln", StringComparison.OrdinalIgnoreCase))
 			{
-				progress.Report("Solution parsing started.");
-				var solution = new SolutionReader().Read(target, progress);
-				if (solution == null)
+				foreach (var project in ConvertSolution(target, conversionOptions, preTransforms, postTransforms, progress))
 				{
-					yield break;
+					yield return project;
 				}
 
-				if (solution.ProjectPaths?.Count > 0)
+				yield break;
+			}
+
+			if (string.Equals(extension, ".csproj", StringComparison.OrdinalIgnoreCase))
+			{
+				// Process only the given project file
+				yield return ProcessFile(target, null, conversionOptions, preTransforms, postTransforms, progress);
+
+				yield break;
+			}
+
+			// Process the only solution in given directory
+			var solutionFiles = Directory.EnumerateFiles(target, "*.sln", SearchOption.TopDirectoryOnly).ToArray();
+			if (solutionFiles.Length == 1)
+			{
+				foreach (var project in ConvertSolution(solutionFiles[0], conversionOptions, preTransforms, postTransforms, progress))
 				{
-					foreach (var projectPath in solution.ProjectPaths)
-					{
-						progress.Report("Project found: " + projectPath);
-						var fullPath = Path.Combine(solution.SolutionFolder.FullName, projectPath);
-						if (!File.Exists(fullPath))
-						{
-							progress.Report("Project file not found at: " + fullPath);
-						}
-						else
-						{
-							yield return ProcessFile(fullPath, solution, conversionOptions, preTransforms, postTransforms, progress);
-						}
-
-					}
+					yield return project;
 				}
 
 				yield break;
 			}
 
 			// Process all csprojs found in given directory
-			if (!Path.GetExtension(target).Equals(".csproj", StringComparison.OrdinalIgnoreCase))
+			var projectFiles = Directory.EnumerateFiles(target, "*.csproj", SearchOption.AllDirectories).ToArray();
```

Comment:
Given the above change to process single solutions, feels like maybe this should get caught up in that too. Let's say the user has a folder with just 10 sub-folders each with 1 solution in. I think it could be done with recursion over the previous conditions.

It messes with my head a bit to think what the most logical thing is though. What you've currently gone for feels a bit more inconsistent and unpredictable than it was before. The option is there to process a solution if that's what people want to do.

Reply:
Maybe an option could be introduced 'process-orphan-projects' or something like that

Reply:
I have always wondered why it would recursively scan for projects. I was surprised when I ran csproj-to-2017 half a year ago in a directory with a solution and a lot of subprojects not intended to be converted. I believe we should respect solution file more and blunt recursive tree descent less.

Reply:
I think the origin of that is the natural progression of:
1. pass me a project file (too much work specifying individual ones)
2. pass me a directory (recursively search projects which would work for a solution)
3. only at this point solution support was added
I still think that the directory support is useful, but that comes down to taste. It did become less useful when solution support was added.